nexus/src/app/background/tasks/support_bundle_collector.rs

```diff
@@ -722,152 +806,322 @@ impl BundleCollection {
+        let status = match self
+            .save_sp_ereports(ereport_filters.clone(), ereports_dir.clone())
             .await
-            {
-                error!(log, "Failed to write sled_info.json"; "error" => InlineErrorChain::new(e.as_ref()));
+        {
+            Ok(n_collected) => {
+                SupportBundleEreportStatus::Collected { n_collected }
             }
+            Err((n_collected, err)) => {
+                warn!(
+                    &self.log,
+                    "Support bundle: sp ereport collection failed \
+                     ({n_collected} collected successfully)";
+                    InlineErrorChain::new(err.as_ref()),
+                );
 
-            let sp_dumps_dir = dir.path().join("sp_task_dumps");
-            tokio::fs::create_dir_all(&sp_dumps_dir).await.with_context(
-                || {
-                    format!(
-                        "Failed to create SP task dump directory {sp_dumps_dir}"
-                    )
-                },
-            )?;
+                SupportBundleEreportStatus::Failed {
+                    n_collected,
+                    error: err.to_string(),
+                }
+            }
+        };
 
-            if let Err(e) =
-                save_all_sp_dumps(log, &mgs_client, &sp_dumps_dir).await
-            {
-                error!(log, "Failed to capture SP task dumps"; "error" => InlineErrorChain::new(e.as_ref()));
-            } else {
-                report.listed_sps = true;
-            };
-        } else {
-            warn!(log, "No MGS client, skipping SP task dump collection");
-        }
+        Ok(CollectionStepOutput::SpEreports(status))
+    }
 
-        if let Ok(all_sleds) = all_sleds {
-            report.listed_in_service_sleds = true;
+    async fn get_or_initialize_mgs_client<'a>(
+        &self,
+        mgs_client: &'a OnceCell<Arc<Option<MgsClient>>>,
+    ) -> &'a Arc<Option<MgsClient>> {
+        mgs_client
+            .get_or_init(|| async {
+                Arc::new(self.create_mgs_client().await.ok())
+            })
+            .await
+    }
 
-            const MAX_CONCURRENT_SLED_REQUESTS: usize = 16;
-            const FAILURE_MESSAGE: &str =
-                "Failed to fully collect support bundle info from sled";
-            let mut set = ParallelTaskSet::new_with_parallelism(
-                MAX_CONCURRENT_SLED_REQUESTS,
+    async fn get_or_initialize_all_sleds<'a>(
+        &self,
+        all_sleds: &'a OnceCell<Arc<Option<Vec<Sled>>>>,
+    ) -> &'a Arc<Option<Vec<Sled>>> {
+        all_sleds
+            .get_or_init(|| async {
+                Arc::new(
+                    self.datastore
+                        .sled_list_all_batched(
+                            &self.opctx,
+                            SledFilter::InService,
+                        )
+                        .await
+                        .ok(),
+                )
+            })
+            .await
```

Comment:
I'm using lazily-initialized variables for "data/clients that might get created as a part of bundle collection, but might not". 

This becomes more relevant with https://github.com/oxidecomputer/omicron/pull/9254, when we may or may not need these values to get initialized at all.